### ISSUE TITLE
Serious Correlations Bug:  Exponential Explosions of Correlations when Objects contain duplicative IOCs in Multiple Events

### ISSUE THREAD

**Issue by packet-rat** [MISP/MISP#7531]
### Work environment

| Questions                 | Answers
|---------------------------|--------------------
| Type of issue             | Bug
| OS version (server)       | Debian, Ubuntu, CentOS, RedHat, ...
| OS version (client)       | MacOS, Win10, Ubuntu, ...
| PHP version               | 7.1
| MISP version / git hash   | 2.4.143, hash of the commit
| Browser                   | If applicable

### Expected behavior

Events with (1) multiple MISP Objects that (2) each contain multiple duplicative actionable Attributes - 
Should Only form [ONE] correlation to other Events with similar conditions.  

For example: 

If I have Ten events which contain 100 botnet Victim/Controller Pairs Objects with the Same Botnet Controller IP, then there should only be Ten Correlations.  **Not 100,000s to 1,000,000s.**

### Actual behavior

Exponential Correlations are formed between EACH duplicate Attribute within an Object to EACH similarly duplicated Attribue embedded within Objects in another Event

### Steps to reproduce the behavior

The following two Conditions lead to exponential 'bad Correlations in MISP:

(1) MISP Events with Multiple MISP Objects that have highly duplicative Attributes

(2) Multiple MISP Events containing (1) MISP Events with Multiple MISP Objects that have highly duplicative Attributes

**Comment by github-germ** [comment]
We've experienced this also big time. Open to suggestions to avoid. Thanks.

**Comment by chrisinmtown** [comment]
Here's example data to give context to a proposal below.

Event E1 has:

    ObjectAttribute oa1
        Attribute a1, ip addr, 1.2.3.4
        Attribute a2, sha code, bbb777
    ObjectAttribute oa2
        Attribute a3, ip addr, 1.2.3.4
        Attribute a4, sha code, ccc888

Event E2 has:

    ObjectAttribute oa3
        Attribute a5, ip addr, 1.2.3.4
        Attribute a6, sha code, ddd999
    ObjectAttribute oa4
        Attribute a7, ip addr, 1.2.3.4
        Attribute a8, sha code, eee000

I believe the MISP correlation machinery creates these links from E1 to E2:

    oa1/a1 -> oa3/a5
    oa1/a1-> oa4/a7
    oa2/a3 -> oa3/a5
    oa2/a3 -> oa4/a7

And similarly for E2 back to E1:

    oa3/a5 -> oa1/a1
    oa3/a5 -> oa2/a3
    oa4/a7 -> oa1/a1
    oa7/a7 -> oa2/a3

We checked to make certain that the correlation table never links an event to itself, for example a correlation oa1/a1 -> oa2/a3, and didn't find any, so I think that's not a concern.

It seems to me that what the correlation engine needs to do when creating correlations is check the event and the attribute value but skip the object-attribute in the middle.  IMHO it's a bug for the correlation engine to consider separately the repeated attribute-value occurrences in different object attributes.  

Using my example above, after the corr. eng. creates the first link oa1/a1->oa3/a5 and considers creating the second (third, fourth) links like oa1/a1->oa4/a7, the correlation feature should detect that an attribute with this value already exists somewhere within E1 and *not* *create* *additional* *links*. 

Stated differently, I'm proposing that the MISP correlation engine should stop super-correlating two events by creating tens, hundreds, thousands, .. of same-value attribute links between the two events. Such a change will stop the correlation explosion problem. Will that change break other assumptions and features?

**Comment by RichieB2B** [comment]
The proposal by @chrisinmtown makes sense to me. We recently ingested a few misp events each with several file objects including `imphash` attributes. These are all equal by design. This caused our correlations table to explode to 7GB on disk (24M correlations).

**Comment by RichieB2B** [comment]
Is there a way to prioritize this fix? I imagine the entire MISP community is running out of disk space by now.

**Comment by iglocska** [comment]
This is definitely not an easy "fix" as it's working as intended in this case. The correlations always happen on an attribute level, never on an object or event level, those are completely ignored (except for self references to the same event, which would be pointless). If we did restrict the number of correlations between individual re-occuring values in an event to a single correlation the result would be:

1. Different distribution values might hide valid correlations from certain users.

oa1/a1 -> oa3/a5
oa1/a1-> oa4/a7
oa2/a3 -> oa3/a5
oa2/a3 -> oa4/a7

In this example, restricting the correlations to just:

oa1/a1 -> oa3/a5

would mean that if oa1/a1 or oa3/a5 has a more restrictive distribution setting attached than say oa2/a3, than users would never see the correlations that are excluded by the more strict distribution setting.

2. Besides that we have another issue, whilst we often view events as is with all the object and attribute relations contained within, for many applications a flat list of attributes is more useful. This is when we use the /attributes/restSearch endpoint to collect our filtered data-set, which we often also use to include correlations for each of the results returned in our query. These would no longer be complete if we restricted the correlations.

So tl;dr: What can I do to avoid this situation then? When ingesting data / creating your ingestion scripts, make sure that you use the flag controlling whether an attribute correlates with care. If you know that you will be describing the same file in 50 different ways in an event, each including the same hashes, consider only setting the correlation flag on a single instance of the attribute (with the above mentioned limitations on distribution settings in mind, so if you decide to use different settings, make sure this is reflected in the attributes that you choose to correlate on). 

Hope this clears it up somewhat.

**Comment by RichieB2B** [comment]
I understand the challenges described by @iglocska and have been thinking on how to solve this. It would be great if the authors of events would realise this problem exists but IMHO that is just wishful thinking.

Our research shows that the recent explosion of the correlations table is caused by the `imphash` type attributes that are by design the same for many files described in file objects. How about automating this process:
If all of these conditions are met:

1. the value of multiple `imphash` attributes is the same
1. the attributes belong to the same event
1. the distribution of the attributes is the same
1. the distribution of the objects the attributes belong to is the same

then automatically set the `disable_correlation` flag on all of them except one.

Is that feasible to automate?

**Comment by github-germ** [comment]
@RichieB2B -- Appreciate the continued discussion here. Perhaps I'm missing something. Might setting a solution based on one attribute type, `imphash`, be too narrow? E.g. our explosion occurs within events with objects containing `ip-src` attributes.

Perhaps there is a way to generalize your proposal with some method defining a list of `MISPObjectAttributes` within an event that should somehow automagically be deduped into `MISPAttributes` within the event (i.e. not within an object) and those can have correlations enabled without setting off the bomb. Or encapsulate that list of those deduped `MISPObjectAttributes` into one object in the event (meta correlation if you will).

**Comment by RichieB2B** [comment]
@github-germ Absolutely, if you are experiencing this issue with `ip-src` we should be looking for a general solution instead of focusing on a single attribute type like `imphash`.

**Comment by packet-rat** [comment]
This is a global issue that spans multiple Attribute types (any Attribute Type that has Correlation enable).
This is a **"Show-Stopper"** issue for those of us trying to leverage and expand our use of MISP Objects.

**Comment by github-germ** [comment]
Is this on the radar?

**Comment by cybgit** [comment]
We've definitely met this issue ourselves with an explosion of correlations due to multiple similar attributes in objects within an event. For example, lots of domain|ip objects that have a different domain but the same IP cause this to happen.

I would love a solution to this. Could this be a toggle/setting that the admin choses? So the fix described, but applied if the admin sets a misp setting flag. something like misp.disable_multiple_correlations_on_objects  (or something less wordy).?

**Comment by packet-rat** [comment]
Looking to re-engage in discourse on this issue. Our perspective is that an Event with an actionable Attribute in an Object (i.e., the C2 Server IP Address in multiple Objects describing Botnet C2/Victim Pairs) **should not Correlate to itself**.   The C2 IP, although repeated, is the most actionable Attribute in this type of data model and we need correlations to other MISP Events.

**Comment by ajjonen** [comment]
Could one use an alternate attribute for the ips not being correlated in the object?  Then set that attribute to not correlate.

**Comment by ajjonen** [comment]
use a text attribute for the 'pairs' in your object and set it to not correlate.  Then use an ip-dst as an attribute of the c2 in that object.
